dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/AssignmentInlining.java

```diff
@@ -0,0 +1,115 @@
+package com.dat3m.dartagnan.program.processing;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.processing.ExprTransformer;
+import com.dat3m.dartagnan.program.Function;
+import com.dat3m.dartagnan.program.Program;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.analysis.DominatorAnalysis;
+import com.dat3m.dartagnan.program.event.Event;
+import com.dat3m.dartagnan.program.event.EventVisitor;
+import com.dat3m.dartagnan.program.event.RegReader;
+import com.dat3m.dartagnan.program.event.RegWriter;
+import com.dat3m.dartagnan.program.event.core.Local;
+import com.dat3m.dartagnan.utils.DominatorTree;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+    This pass performs the following inlining of assignments:
+
+          r = someExpr;      =====>       r = someExpr;
+          RegReader(r);      =====>       RegReader(someExpr); // uses "someExpr" directly
+
+    The inlined assignments are dead and can be eliminated by dead assignment eliminations.
+    To avoid duplication of "someExpr", the pass only inlines assignments if the assigned register is used only once.
+    To achieve this, the pass scans the code twice: once to collect usage statistics and once to perform the inlining.
+ */
+public class AssignmentInlining implements FunctionProcessor {
+
+    private AssignmentInlining() { }
+
+    public static AssignmentInlining newInstance() {
+        return new AssignmentInlining();
+    }
+
+    @Override
+    public void run(Function function) {
+        if (!function.hasBody() || function.getProgram().getFormat() == Program.SourceLanguage.LITMUS ) {
```

Comment:
This looks like a hack.

What is the problem? I assume it has something to do with final assertions. If that if the case, we should make the pass work with those (we will eventually need this for spirv) instead of skipping it

Reply:
Both this pass and SCCP are sound for litmus tests (and litmus-style assertions). However, we deliberately do avoid optimizations on litmus code and that's why we added the NOOPT tag.
It is currently easier to just skip over litmus than to check for the NOOPT tag all the time.
In fact, it might be even better to have a separate processing pipeline for litmus code that does not even have these passes in the first place.

If we decide to optimize litmus code then we should first get rid of the NOOPT tag or put a more clear semantics to it (i.e., not blindly place it over the whole code). Updating the passes is then just a matter of removing the litmus check and possibly adding special-case handling of NOOPT tags if we still need them for some reason.

Reply:
> However, we deliberately do avoid optimizations on litmus code and that's why we added the NOOPT tag. 

It is not true the NOOPT tag is just about litmus tests. We also need this for e.g. fake dependencies and spin-termination events.

> It is currently easier to just skip over litmus than to check for the NOOPT tag all the time. In fact, it might be even better to have a separate processing pipeline for litmus code that does not even have these passes in the first place.

But then instead of choosing to disable just SCCP and AssignmentInlining, we could check the format when calling `ProcessingManager.run`. If we do so, we can remove the part that tags every litmus event as NOOPT and check if every other occurrence of NOOPT is indeed necessary.

Reply:
> > However, we deliberately do avoid optimizations on litmus code and that's why we added the NOOPT tag.
> 
> It is not true the NOOPT tag is just about litmus tests. We also need this for e.g. fake dependencies and spin-termination events.

Ok, it is true that there are other use-cases. While I think most if not all of them can be removed (e.g., fake dependencies are preserved anyways by the simplifier), if we want to support NOOPT for arbitrary events, I have to add support for this. I think most passes ignore NOOPT currently, so they all need fixing.